src/Engine/Utils.java

```diff
@@ -92,13 +92,14 @@ public static ArrayList<XYCoord> findUnloadLocations(GameMap map, Unit transport
   {
     ArrayList<XYCoord> locations = findLocationsInRange(map, moveLoc, 1);
     ArrayList<XYCoord> dropoffLocations = new ArrayList<XYCoord>();
-    final MoveType cargoMoveType = new UnitContext(cargo).calculateMoveType();
-    if( cargoMoveType.canTraverse(map.getEnvironment(moveLoc)) )
+    final FloodFillFunctor cargoMoveType = new UnitContext(cargo).calculateMoveType().getUnitMoveFunctor(cargo, false, false);
+    if( MoveType.IMPASSABLE > cargoMoveType.getTransitionCost(map, moveLoc, moveLoc) )
```

Comment:
This new code is not as idiomatic as what it replaces.
I'm also not sure why this change is needed? We already couldn't drop across invalid transitions (e.g. a lander dropping an inf onto land via a cliff).
Also don't like having two different classes responsible for movement. Is there a way to use `MoveType` as the interface and have it use `FloodFillFunctor` under the hood to simplify usage?

Reply:
You're right.
The issue I was fixing is "You can unload units onto Teletiles"
The meat of this change is in the `cargoMoveType.canEnd(map, loc)` check below. I looked at the logic of `canTraverse()`, and it doesn't check whether it's a valid tile to end a move on.... and I assumed I shouldn't touch that.
`FFF.canEnd()` checks explicitly for Teletile and for there being a unit present.
Looking at the current usecases for `canTraverse()` (Wally's "usable area", teleport/create's valid destination check, unload), I think it's a good idea to add `canEnd()`'s logic in there.

I'm thinking maybe `MoveType` can function as an FFF with `includeOccupied` and `canTravelThroughEnemies` set to false?
Or maybe there's a better idiom for how we handle `includeOccupied` and `canTravelThroughEnemies`, like having them as optional parameters for `MoveType` and removing FFF entirely.

I guess for this PR, I'll shuffle that logic into `canTraverse()` and then tackle that refactor later? Or I can just do it here, probably.

Reply:
The new change is indeed more idiomatic, but it unfortunately prevents all movement across teletiles. [Hah](https://github.com/ThislsAUsername/DefendPeace/pull/377#issue-1569182892)

Reply:
I absolutely did test it, though - I don't see this bug on my end.

Reply:
Huh. On second inspection, you are correct. I think I am guilty of attempting code review after midnight.
Though in my defense, a plain and simple reading of MoveType.java lends itself to that interpretation:
```java
  /** Returns whether the unit can travel in the specified environment. */
  public boolean canTraverse(Environment tile)
  {
    if( tile.terrainType == TerrainType.TELETILE )
      return false;
    return getMoveCost(tile) < IMPASSABLE;
  }
```
Which makes me wonder just what this code is doing, if it doesn't actually relate to a unit's ability to traverse through tiles. Things seem to have gotten muddy somewhere.

Reply:
`canTraverse()` has basically been the check used for "can I put this unit here?". The movement calc code doesn't use it because it needs to handle the precise movecost anyway.
So... yeah, I'd agree it's a little muddy.
Should I rename it to `canEnd()` to match the FFF's name for the concept?

Reply:
Yes, or maybe `canStandOn()` or `canEndMovement/TurnOn()` or something that will be less misleading.